pkg/dockerregistry/server/tagservice.go

```diff
@@ -0,0 +1,208 @@
+	dgst, err := digest.ParseDigest(te.Image)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+
+	if t.pullthrough {
+		image, err := t.getImage(dgst)
+		if err != nil {
+			return distribution.Descriptor{}, err
+		}
+
+		if !isImageManaged(image) {
+			return distribution.Descriptor{}, distribution.ErrTagUnknown{Tag: tag}
+		}
+	}
+
+	return distribution.Descriptor{Digest: dgst}, nil
+}
+
+func (t tagService) All(ctx context.Context) ([]string, error) {
+	tags := []string{}
+
+	imageStream, err := t.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.namespace, t.name, err)
+		return tags, distribution.ErrRepositoryUnknown{Name: t.repository.Named().Name()}
+	}
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+		if t.pullthrough {
+			dgst, err := digest.ParseDigest(history.Items[0].Image)
+			if err != nil {
+				return nil, err
+			}
+
+			image, err := t.getImage(dgst)
+			if err != nil {
+				return nil, err
+			}
+
+			if !isImageManaged(image) {
+				continue
+			}
+		}
+		tags = append(tags, tag)
+	}
+	return tags, nil
+}
+
+func (t tagService) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
+	tags := []string{}
+
+	imageStream, err := t.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.namespace, t.name, err)
+		return tags, distribution.ErrRepositoryUnknown{Name: t.repository.Named().Name()}
+	}
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+
+		dgst, err := digest.ParseDigest(history.Items[0].Image)
+		if err != nil {
+			return tags, err
+		}
+
+		if t.pullthrough {
+			image, err := t.getImage(dgst)
+			if err != nil {
+				return nil, err
+			}
+
+			if !isImageManaged(image) {
+				continue
+			}
+		}
+
+		if dgst == desc.Digest {
+			tags = append(tags, tag)
+		}
+	}
+
+	return tags, nil
+}
+
+func (t tagService) Tag(ctx context.Context, tag string, dgst distribution.Descriptor) error {
+	imageStream, err := t.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.namespace, t.name, err)
+		return distribution.ErrRepositoryUnknown{Name: t.repository.Named().Name()}
+	}
+
+	image, err := t.registryClient.Images().Get(dgst.Digest.String())
+	if err != nil {
+		context.GetLogger(ctx).Errorf("(tagService).Tag: Unable to get image: %s", dgst.Digest.String())
+		return err
+	}
+	image.SetResourceVersion("")
+
+	if t.pullthrough && !isImageManaged(image) {
+		return distribution.ErrRepositoryUnknown{Name: t.repository.Named().Name()}
+	}
+
+	ism := imageapi.ImageStreamMapping{
+		ObjectMeta: kapi.ObjectMeta{
+			Namespace: imageStream.Namespace,
+			Name:      imageStream.Name,
+		},
+		Tag:   tag,
+		Image: *image,
+	}
+
+	return t.registryClient.ImageStreamMappings(imageStream.Namespace).Create(&ism)
```

Comment:
Since we already use `return t.registryClient.ImageStreamTags` I think it's ok to switch to `ImageStreamTags.Update()` which is more natural.